### PR TITLE
Generating text from Indonesian Wikipedia LM

### LINKED ISSUE #4: Generate text using Indo LM
Write a predict function which basically generates K tokens given a start seed. Something to the effect of:
```
lm.predict(next_tok:int =K, start="Saya Meghana, Insinyur Pembelajaran dan Bahasa Dalam")
> # your next K token go here ...
```

### PR DESCRIPTION
**Pull request by meghanabhange**
- Fixes #4
- Generates text given number of tokens and seed start
- Training Scripts for Wikipedia LM

### CODE REVIEW

File: src/utils.py [hunk 1]
```diff
@@ -20,23 +23,43 @@ def __init__(self, data_path):
...
-    def downloader(self, download_type):
+    def downloader(self, download_file_name):
         """
-        Downloads download_type in data_path/download_type.
+        Downloads download_file_name in data_path/download_file_name.
         """
-        download_logger.info(f"Downloading {download_type}")
-        fname = f"{download_type}.tar.gz"
+        download_logger.info(f"Downloading {download_file_name}")
+        fname = f"{download_file_name}.tar.gz"
```
Comment (NirantK) [inline src/utils.py:25]: So glad this `download_file_name`. I was waiting to see how long it takes you to get this fixed :P

File: src/utils.py [hunk 2]
```diff
@@ -20,23 +23,43 @@ def __init__(self, data_path):
...
+        if not (self.data_path / f"{download_file_name}").exists():
             output_unzip = subprocess.check_output(
                 ["tar", "xvzf", f"{self.data_path/fname}", "-C", f"{self.data_path}"]
             )
             download_logger.info(f"{output_unzip}")
-        download_logger.info(f"{download_type} Downloaded")
+        download_logger.info(f"{download_file_name} Downloaded")
+
+
+class LangTokenizer(BaseTokenizer):
```
Comment (NirantK) [inline src/utils.py:43]: `LangTokenizer` class should be imported and used as is from fastai. It already has a lang param IIRC.

File: src/wiki_ulmfit.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+logger = logging.getLogger()
+
+
+class WikiTrainer:
+    def __init__(self, path, lang="id"):
+        self.path = path
+        self.lang = lang
+
+    def download_wiki_files(self, wiki_text=False):
```
Comment (NirantK) [inline src/wiki_ulmfit.py]: Never mix a download function with your Trainer wheels.

File: src/wiki_ulmfit.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        """
+        Loads data_lm for wikipedia data.
+        """
+        tokenizer = Tokenizer(tok_func=LangTokenizer, lang=self.lang)
+        if (self.path / "data_save.pkl").exists():
+            data_lm = load_data(self.path)
+        else:
+            data_lm = TextLMDataBunch.from_csv(
+                self.path,
+                f"{self.lang}_wiki.txt",
```
Comment (NirantK) [inline src/wiki_ulmfit.py]: Paramterize the filename as an arg of the function or ideally, the class init. This way, you can do the `exists()` check for all necessary ingredients in one place. 

And not waste time down the pipeline if something is broken.

File: src/wiki_ulmfit.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        learn = language_model_learner(self.data_lm, AWD_LSTM, drop_mult=0.3)
+        if encoder:
+            learn.load_encoder(model_name)
+        else:
+            learn.load(model_name)
+        output_text = learn.predict(start, next_tok)
+        return output_text
+
+
+def main():
```
Comment (NirantK) [inline src/wiki_ulmfit.py:163]: This should be `Fire`d

File: src/wiki_ulmfit.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                f"{self.lang}_wiki.txt",
+                tokenizer=tokenizer,
+                bs=48,
+                text_cols=0,
+                label_cols=None,
+            )
+            data_lm.save()
+        self.data_lm = data_lm
+
+    def train(self, lr=5e-4, n_epocs=1, fname_out="idwiki-latest"):
```
Comment (NirantK) [inline src/wiki_ulmfit.py]: Separate the training config from training code into something else. 

And might be worthwhile to keep a consistent API convention. 
E.g. 
```
wikilearner = WikiLearn(**params)
wikilearner.fit()
```

File: README.md [hunk 7]
```diff
@@ -5,5 +5,22 @@ Summarisation Using ULMFiT Hopefully?
...
+    python3 -m src.download --filename indosum
+    python3 -m src.download --filename indo_lm
+    python3 -m src.download --filename sentencepiece
 ```
+
+## wiki_ulmfit
+
+```python
+    wiki_trainer = WikiTrainer(Path("./data"))
+    wiki_trainer.download_wiki_files()
```
Comment (NirantK) [inline README.md]: Download should precede object initialisation, so that we can whether all dependencies are valid and there during object init.

File: src/wiki_ulmfit.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                lang_dict[self.lang]["data"], f"{self.path}/data_save.pkl", quiet=False
+            )
+        if wiki_text:
+            gdown.download(
+                lang_dict[self.lang["wiki"]],
+                f"{self.path}/{self.lang}_wiki.txt",
+                quiet=False,
+            )
+
+    def load_data_lm(self):
```
Comment (NirantK) [inline src/wiki_ulmfit.py]: How does the API change if I want to use a finetuned LM instead of Wiki LM to predict text? 

Decouple the train, load and predict functions form "Wiki" models. It should be able to work with any fastai LM.

File: src/wiki_ulmfit.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        
+        Returns:
+            [str] -- Ouput predicted string.
+       """
+        learn = language_model_learner(self.data_lm, AWD_LSTM, drop_mult=0.3)
+        if encoder:
+            learn.load_encoder(model_name)
+        else:
+            learn.load(model_name)
+        output_text = learn.predict(start, next_tok)
```
Comment (NirantK) [inline src/wiki_ulmfit.py]: `learn.predict` takes more args, accept them as **kwargs or similar

**Review by NirantK — CHANGES_REQUESTED** [review]
Broadly, good job on the code. I am assuming it works. 

Just a main edge case: Make sure it works with any LM. 

Your classes for load, predict specially. And they should not be tied into train in anyway.

File: src/wiki_ulmfit.py (hunk 3 above)
Comment (meghanabhange) [reply src/wiki_ulmfit.py]: Moved download_wiki_files to download.py

File: src/utils.py (hunk 1 above)
Comment (meghanabhange) [reply src/utils.py:25]: Well--don't know what to say there :P

File: src/wiki_ulmfit.py (hunk 8 above)
Comment (meghanabhange) [reply src/wiki_ulmfit.py]: Created `load_data_file` instead which will load any TextBunch given either the pickle file name or `.txt` file.

File: src/utils.py [hunk 10]
```diff
@@ -57,3 +58,29 @@ def tokenizer(self, t: str) -> List[str]:
...
+
+    def on_epoch_end(self, epoch, **kwargs: Any) -> None:
+        "Compare the value monitored to its best score and maybe stop training."
+        current = self.get_monitor_value()
+        if not current:
+            return None
+        print(
+            f"Heyya, This is Current : {current} and expected value {self.expected_value}"
+        )
+        if current <= self.expected_value:
```
Comment (NirantK) [inline src/utils.py]: This is not the meaning of Early Stopping! Early Stopping kicks in when you have not seen any improvement in the network for a few iterations: https://machinelearningmastery.com/how-to-stop-training-deep-neural-networks-at-the-right-time-using-early-stopping/

File: src/wiki_ulmfit.py [hunk 11]
```diff
@@ -79,7 +85,15 @@ def load_language_model(
             fname_text_file=f"{self.lang}_wiki.txt",
             lang=self.lang,
         )
-        learn = language_model_learner(self.data_lm, AWD_LSTM, drop_mult=0.3)
+        learn = language_model_learner(
+            self.data_lm,
+            AWD_LSTM,
+            drop_mult=0.3,
+            metrics=[accuracy, Perplexity()],
```
Comment (NirantK) [inline src/wiki_ulmfit.py:112]: Why is Perplexity both in metrics and callbacks? I don't follow

File: src/utils.py (hunk 10 above)
Comment (meghanabhange) [reply src/utils.py]: Used the right early stopping

File: src/utils.py (hunk 2 above)
Comment (meghanabhange) [reply src/utils.py:43]: Using LangTokenizer while loading databunch
